crates/agent-sql/src/evolutions.rs

```diff
@@ -0,0 +1,183 @@
+    .await?;
+
+    Ok(())
+}
+
+pub struct LiveSpecRow {
+    pub id: Id,
+    pub created_at: DateTime<Utc>,
+    pub updated_at: DateTime<Utc>,
+    pub last_pub_id: Id,
+    pub catalog_name: String,
+    pub spec_type: Option<CatalogType>,
+    pub spec: Option<Json<Box<RawValue>>>,
+    pub spec_capabilities: Json<Vec<RoleGrant>>,
+    pub user_capability: Option<Capability>,
+}
+
+pub async fn fetch_live_specs(
+    user_id: Uuid,
+    catalog_names: Vec<String>,
+    txn: &mut sqlx::Transaction<'_, sqlx::Postgres>,
+) -> sqlx::Result<Vec<LiveSpecRow>> {
+    sqlx::query_as!(
+        LiveSpecRow,
+        r#"
+            select
+                l.id as "id: Id",
+                l.created_at,
+                l.updated_at,
+                l.last_pub_id as "last_pub_id: Id",
+                l.catalog_name,
+                l.spec as "spec: Json<Box<RawValue>>",
+                l.spec_type as "spec_type: CatalogType",
+                coalesce(
+                    (select json_agg(row_to_json(role_grants))
+                    from role_grants
+                    where starts_with(l.catalog_name, subject_role)),
+                    '[]'
+                ) as "spec_capabilities!: Json<Vec<RoleGrant>>",
+                (
+                    select max(capability) from internal.user_roles($2) r
+                    where starts_with(l.catalog_name, r.role_prefix)
+                ) as "user_capability: Capability"
+            from live_specs l
+            where l.catalog_name = ANY($1)
+        "#,
+        catalog_names.as_slice(),
+        user_id,
+    )
+    .fetch_all(txn)
+    .await
+}
+
+pub struct DraftSpecRow {
+    pub draft_spec_id: Id,
+    pub live_spec_id: Option<Id>,
+    pub catalog_name: String,
+    pub expect_pub_id: Option<Id>,
+    pub last_pub_id: Option<Id>,
+    pub draft_spec: Option<Json<Box<RawValue>>>,
+    pub draft_type: Option<CatalogType>,
+    pub live_type: Option<CatalogType>,
+    pub user_capability: Option<Capability>,
+}
+
+pub async fn fetch_draft_specs(
+    draft_id: Id,
+    user_id: Uuid,
+    txn: &mut sqlx::Transaction<'_, sqlx::Postgres>,
+) -> sqlx::Result<Vec<DraftSpecRow>> {
+    sqlx::query_as!(
+        DraftSpecRow,
+        r#"
+        select
+            draft_specs.id as "draft_spec_id: Id",
+            draft_specs.catalog_name,
+            draft_specs.expect_pub_id as "expect_pub_id: Id",
+            draft_specs.spec as "draft_spec: Json<Box<RawValue>>",
+            draft_specs.spec_type as "draft_type: CatalogType",
+            live_specs.spec_type as "live_type: CatalogType",
+            live_specs.last_pub_id as "last_pub_id: Option<Id>",
+            live_specs.id as "live_spec_id: Option<Id>",
+            (
+                select max(capability) from internal.user_roles($2) r
+                where starts_with(draft_specs.catalog_name, r.role_prefix)
+            ) as "user_capability: Capability"
+        from draft_specs
+        left join live_specs
```

Comment:
I'd recommend making this a left outer join and filtering `live_specs` to those the user has a `read` capability for. If the user shouldn't see it, why even bring it into the application?

Reply:
> If the user shouldn't see it, why even bring it into the application?

I don't have a _great_ answer to this. Originally, I was thinking we might want to take some other action to notify the owner of that other materialization that one of its bound collections is changing, but that doesn't actually make tons of sense given that we're only operating on a draft. At the moment, the only purpose is so we can log that we're not updating it, which admittedly doesn't seem worth it. I'll just change this to filter them out here.

Reply:
Caught up about this on VC, and the resolution is that there's no longer a need to pull in `user_capability` here since the user has already added the spec to their draft. So they must already have `read` access at a minimum, and authZ issues can get caught when they try to publish.